### DISCUSSION TITLE
Pop camera immediately after image taken

### DISCUSSION THREAD

**Discussion by MadsStolt** [Apparence-io/CamerAwesome#235] [Q&A] [answered]
I dont seem to be able to listen for the camera capture event. In my usecase I want to be able to open the camera capture an image and then pop the camera immediately after (without tapping the preview). However it only seems there is an "onMedieaTap" event. Anyone got a work around for this?

**Comment by apalala-dev** [comment, marked as answer]
You can listen to `cameraState.captureState$` to react to different capture states.
In your case, you probably want a `MediaCapture` with a `MediaCaptureStatus.success` before popping.

**Reply by MadsStolt** [reply]
Exactly what I needed. Thank you kind sir.

**Reply by nathCode45** [reply]
How exactly do you set up such a listener? Any help would be greatly appreciated

**Reply by dumabg** [reply]
The some problem. I have solved it like this:

Define a variable for a subscription:

`StreamSubscription<MediaCapture?>? _cameraStateSubscription;`

Create a `CameraAwesomeBuilder.custom` and capture the stream the first time on the builder property:

```
return CameraAwesomeBuilder.custom(
        builder: (CameraState state, previewSize, previewRect) {
      _cameraStateSubscription ??=
          state.captureState$.listen(controller.onCameraStateChanged);
      return AwesomeCameraLayout(
        state: state,
      );
    },
```

Cancel the subscription on dispose:

```
 @override
  void dispose() {
    _cameraStateSubscription?.cancel();
    super.dispose();
  }
```

**Reply by tiltmaster** [reply]
Hello, do i need to use the CameraAwesomeBuilder.custom ? i am using the CameraAwesomeBuilder.awesome can that work or not?

**Reply by tiltmaster** [reply]
I managed to do it thanks

**Reply by eudaldpons** [reply]
Hi, @tiltmaster can you share your code? thanks in advance

**Reply by Tryanium** [reply]
Hi ! Did you manage? I'm stuck with what the controller is on the previous example?

**Comment by tiltmaster** [comment]
@eudaldpons @Tryanium , well lets start off by defining the following: 

 ` StreamSubscription<MediaCapture?>? _cameraStateSubscription;
  bool isProcessing = false;`

then we must use a `CameraAwesomeBuilder.custom`, and define in the builder as such :

```
CameraAwesomeBuilder.custom(
        builder: (CameraState state, previewSize, previewRect) {
          _cameraStateSubscription ??= state.captureState$.listen(
            (event) {
              if (event?.status == MediaCaptureStatus.capturing) {
                setState(() {
                  isProcessing = true;
                });
              }
              if (event?.status == MediaCaptureStatus.success) {
                event?.captureRequest.when(single: (single) {
                  yourImageFunctionToProcessTheFile(single.file!.path);
                });
              }
            },
          );
          return Container 
// return of container code
```

when using CameraAwesomeBuilder.custom you do need to define the buttons as per the documentation (buttons to take image etc.), since my usecase was to capture one image and then process it and not allowing anything else, this is then is defined in a column (usually covering the entire page)

 ```
AwesomeFilterWidget(state: state, ), //for its required to pass the state of CameraAwesome. and then define the button as such: 
 
IgnorePointer(
  ignoring: isProcessing,
  child: AwesomeCaptureButton(state: state),
 ),
```

with that you should be able to click the button, sets the isProcessing to true making the button not clickable and then since ur listening to `state.captureState$.listen` it will call the function and process the image as needed.

the fact that i let this just to see the Author and or anyone else post a solution to it, but once again both github & stackoverflow disappoint me as usually they never offer such responses to questions.

anyhow, i hope this solves your issue.

**Reply by eudaldpons** [reply]
Thanks mister, i will try it

**Reply by Tryanium** [reply]
Thx Tiltmaster ! I have tried to make it works, unfortunately I have this error while trying

`Function expressions can't be named.
Try removing the name, or moving the function expression to a function declaration statement.dart(named_function_expression)`

On the `.listen` part, and I'm not sure where to put the `AwesomeFilterWidget` part, I'm kinda new to flutter

**Reply by Tryanium** [reply]
This is my whole code

```dart
import 'package:flutter/material.dart';
import 'dart:io';
import 'dart:async';
import 'package:path_provider/path_provider.dart';

import 'package:camerawesome/camerawesome_plugin.dart';

class CameraPage extends StatefulWidget {
  const CameraPage({super.key});

  @override
  State<CameraPage> createState() => _CameraPageState();
}

class _CameraPageState extends State<CameraPage> {
  late String imagePath = '';
  @override
  void dispose() {
    super.dispose();
  }

   StreamSubscription<MediaCapture?>? _cameraStateSubscription; bool isProcessing = false;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        body: CameraAwesomeBuilder.custom(
            saveConfig: SaveConfig.photo(
              pathBuilder: (sensors) async {
                // 1.
                final Directory extDir = await getTemporaryDirectory();
                final testDir = await Directory('${extDir.path}/camerawesome')
                    .create(recursive: true);

                // 2.
                if (sensors.length == 1) {
                  final String filePath =
                      '${testDir.path}/${DateTime.now().millisecondsSinceEpoch}.jpg';
                  // 3.
                  return SingleCaptureRequest(filePath, sensors.first);
                } else {
                  // 4.
                  return MultipleCaptureRequest(
                    {
                      for (final sensor in sensors)
                        sensor:
                            '${testDir.path}/${sensor.position == SensorPosition.front ? 'front_' : "back_"}${DateTime.now().millisecondsSinceEpoch}.jpg',
                    },
                  );
                }
              },
            ),
            previewFit: CameraPreviewFit.fitWidth,
            previewAlignment: Alignment.center,
            enablePhysicalButton: true,
            builder: (CameraState state, previewSize, previewRect) {
              _cameraStateSubscription ??= state.captureState$.listen(event) {
                if(event?.status == MediaCaptureStatus.capturing) {
                  setState(() {
                    isProcessing = true;
                  });
                }
                if(event?.status == MediaCaptureStatus.success) {
                  event?.captureRequest.when(single: (single) {
                    yourImageFunctionTOProcessTheFile(single.file!.path);
                  });
                }
              };
              AwesomeFilterWidget(state: state, ); 
              IgnorePointer(
                ignoring: isProcessing,
                child: AwesomeCaptureButton(state: state),);
              return AwesomeCameraLayout(state: state
              );
              
            },

              ),
            
            );
            
            /*
            topActionsBuilder: (state) =>
                AwesomeTopActionsCustom(state: state)));
                */
  }
}
```

**Reply by linda-mastertech** [reply]
Thank you @Tryanium, your code above worked for me.
